### ISSUE TITLE
Make Unison a portable program under Windows

### ISSUE THREAD

**Issue by stdedos** [bcpierce00/unison#604]
I'd like it so that the `%USERPROFILE%/.unison` folder can be ignored (and not created) if there is such folder "somewhere around" the executable.

While maybe it doesn't make sense for some "state/temporary" files, I'd like to be able to carry `.prf` files cleanly among computers

**Comment by gdt** [comment]
Please bring this up on the users list.   I don't understand exactly what you mean, and more importantly I don't understand why it will be sound.    I also don't understand what this has to do with windows.

**Comment by stdedos** [comment]
> Please bring this up on the users list.

I could, but I don't want to register, get the mailing list's spam, only to track my issue 😕 

> I don't understand exactly what you mean, and more importantly I don't understand why it will be sound.

There is an application concept in Windows called "Portable Apps" (https://portableapps.com/about/what_is_a_portable_app#definition, one of the pages that provides it). 

The artifact you are providing on the Releases is a full, working app - and it has (almost) no dependencies outside of that artifact. It could be called Portable in that sense 

I'd like that it is also possible that, any configuration can also be added on a folder inside that Release artifact structure - so that the app is closer to be fully portable. 

> I also don't understand what this has to do with windows.

"Portable App" is a (considerably big) concept on Windows. I don't know whether it's a concept in Linux/MacOS - and therefore I focus on what I know.

**Comment by gdt** [comment]
I'm sorry that you don't wish to engage with the unison developer community.

I still don't understand how this is supposed to work in terms of synchronizing things that are within the package or things that are not within the package.  The archive roots have to be bound to the data. 

I've marked this wontfix as I don't use Windows and I am not aware of anyone actively contributing who is inclined to add Windows-specific features.  If there's a PR, I think it's pretty likely it will get looked at.

**Comment by stdedos** [comment]
> I still don't understand how this is supposed to work in terms of synchronizing things that are within the package or things that are not within the package. The archive roots have to be bound to the data.

I think it's orthogonal to that?

tl;dr I wish that this would work:
![image](https://user-images.githubusercontent.com/133706/142742630-e25a6c51-1222-40f1-82b5-311c173fa1d6.png)

_Instead of_ saving the files to `%USERPROFILE%/.unison`

**Comment by acolomb** [comment]
I don't think you would gain much by keeping the directory "portable".  Unison is just designed to be bound to a single computer / user per instance.  Obviously the root paths in the profiles are not necessarily portable.  Then the data itself needs to be ported as well.  And finally each of the archive files encodes the hostname of the machine where it was created; a different host will lead to different archive file names.  So all in all you will have much bigger troubles even if you could manage to keep `.unison` along with the binary.  That might be desirable for some reasons (although in general the trend is to separate the generic program code from user-specific data and settings), but certainly not useful to make the Unison installation "portable" (as in reusable over a network share or to plug a pen drive into different computers).

If you want to share the profile definitions, I'd advise to sync all `*.prf` and `*.inc` files to a shared folder from each machine's user-specific folder.  You can even use Unison itself for that task.  If you are careful that the root paths look the same everywhere, it works pretty well.  I am doing it as well, mainly to have them readily accessible as reference but also to keep Unison settings in sync across different hosts.

**Comment by stdedos** [comment]
> That might be desirable for some reasons (although in general the trend is to separate the generic program code from user-specific data and settings), ...

There are other setups, e.g., my data drive and my OS/`%USERPROFILE%` folder are in separate physical devices - therefore preserving application data "should" be transparent even to an OS installation (ofc there's the `hostname`, but that's okay).

> ... but certainly not useful to make the Unison installation "portable" (as in reusable over a network share or to plug a pen drive into different computers).

Syncing a remote path to the user's downloads folder (by getting the path **correctly** via https://stackoverflow.com/a/48706260, or just `%USERPROFILE%/Downloads`) seems pretty straight-forward to me as portable-possible.

> I'd advise to sync all *.prf and *.inc files to a shared folder from each machine's user-specific folder.

Idea seems promising, but it still "defeats" the purpose: `mv downloads.{prf,inc}`, means that `echo "include $(realpath ${SYNC_PATH}/downloads.inc)" > downloads.prf` (or Windows-equivalent) is at least necessary to get Unison working again

**Comment by gdt** [comment]
I still haven't seen an explanation of what would be synced and how (soundly), with software and config dir on an external USB, that's hard to do with command-line args now.   Please feel free to discuss on unison-hackers.  Or, submit a pull request that also adjusts docs and explains concept of operations, but I see that as pushing a wet noodle uphilll, at least until it can be explained how it will be used and why that makes sense.

**Comment by stdedos** [comment]
> I still haven't seen an explanation of what would be synced and how (soundly), with software and config dir on an external USB, that's hard to do with command-line args now.

The .prf files. The configuration. I know that "theoretically" everything should be configurable via the cli as well; however, you cannot argue that "cli is easier", especially when we are talking about big .prf files. (especially on windows terminal).
I don't think it's a matter of "what" would be so complex that the .prf files would be **mandatory** - it's just preference to sync configuration-as-a-file instead of configuration-as-a-cmd spaghetti.

> Please feel free to discuss on unison-hackers.

I have nothing against any mail list; but my mail is already busy/spammed enough. I could bear with the traffic I cause; I don't want the rest of the mailing list "coming my way" too. The forum is nice enough here I'd say.

> ... how it will be used and why that makes sense.

If you cannot understand what a Portable program in Windows is with the above explanation (or the https://docs.appimage.org/user-guide/portable-mode.html which is described from the Linux perspective), then I cannot help more - I am not sure how can I come through and be more explanative. 
If you don't/cannot see the appeal of it, then **most definitely** a PR might not/will not change this - let alone the extensiveness that you request on a change that don't foresee to be more than 10-20 loc.

**Comment by tleedjarv** [comment]
This has always been possible. Just set the `UNISON` environment variable which can contain a relative or absolute path. The value of this variable will be used as the `.unison` dir.

In your case, you could for example set the `UNISON` env var to just `.unison` (or even `profiles_and_archives`) and it will be relative to where `unison.exe` is started.

**Comment by stdedos** [comment]
Wooow 🤩 So simple in the end!

Thank you!
